daisy_workflows/image_import/suse/translate.py

```diff
@@ -26,67 +26,159 @@
+    except Exception as e:
+      if pkg.required:
+        raise AssertionError(
+            'Failed to install required package {}: {}'.format(pkg.name, e))
+      else:
+        logging.warning(
+            'Failed to install optional package {}: {}'.format(pkg.name, e))
 
 
-def DistroSpecific(g):
-  install_gce = utils.GetMetadataAttribute('install_gce_packages')
+def update_grub(g):
+  g.command(
+      ['sed', '-i',
+       r's#^\(GRUB_CMDLINE_LINUX=".*\)"$#\1 console=ttyS0,38400n8"#',
+       '/etc/default/grub'])
+  g.command(['grub2-mkconfig', '-o', '/boot/grub2/grub.cfg'])
 
-  # Remove any hard coded DNS settings in resolvconf.
-  logging.info('Resetting resolvconf base.')
-  g.sh('echo "" > /etc/resolv.conf')
 
-  # Try to reset the network to DHCP.
-  g.write('/etc/sysconfig/network/ifcfg-eth0', network)
+def reset_network(g):
+  logging.info('Updating network to use DHCP.')
+  g.sh('echo "" > /etc/resolv.conf')
+  g.write('/etc/sysconfig/network/ifcfg-eth0', '\n'.join((
+      'BOOTPROTO=dhcp',
+      'STARTMODE=auto',
+      'DHCLIENT_SET_HOSTNAME=yes')
+  ))
 
-  if install_gce == 'true':
-    g.command(['zypper', 'refresh'])
 
-    logging.info('Installing cloud-init.')
-    g.command(['zypper', '-n', 'install', '--no-recommends', 'cloud-init'])
+def install_gcp_licensing(distro, g):
+  if distro.flavor == 'opensuse':
+    return
+  licensing = utils.GetMetadataAttribute('licensing')
+  if not licensing:
+    raise AssertionError(
+        'licensing attribute must be set to either gcp or byol.')
+  if licensing == 'byol':
+    return
 
-    # Installing google-compute-engine-init and not installing
-    # gce-compute-image-packages as there is no port to SUSE
-    logging.info('Installing GCE packages.')
-    g.command(
-        ['zypper', '-n', 'install', '--no-recommends',
-        'google-compute-engine-init'])
+  # TODO: Implement GCP licensing for SLES.
+  # Here are SUSE's instructions:
+  #   https://www.suse.com/c/create-sles-demand-images-gce/
 
-    logging.info('Enable google services.')
-    g.sh('systemctl enable /usr/lib/systemd/system/google-*')
+  raise AssertionError(
+      'GCP licensing not supported for imported SLES images. Please use BYOL.')
 
-    # Try to install the google-cloud-sdk package. It may be not available on
-    # all Leap versions so don't raise an error if it fails.
-    try:
-      g.command(
-          ['zypper', '-n', 'install', '--no-recommends',
-          'google-cloud-sdk'])
-    except Exception as e:
-      logging.warn("Optional google-cloud-sdk package couldn't be installed: "
-                   "%s" % e)
 
-  # Update grub config to log to console, remove quiet and timeouts.
-  g.command(
-      ['sed', '-i',
-       '-e', r'/GRUB_CMDLINE_LINUX/s#"$# console=ttyS0,38400n8'
-       ' scsi_mod.use_blk_mq=Y"#',
-       '-e', r'/GRUB_CMDLINE_LINUX/s#quiet##',
-       '-e', r'/GRUB_[^=]*TIMEOUT=/s#=.*#=0',
-       '/etc/default/grub'])
+def install_virtio_drivers(g):
```

Comment:
This is interesting. This shouldn't be necessary- did you run into problems here?

Reply:
I ran into a problem with SLES 12.5 image that I created with vmware. SLES's installer didn't include virtio kernel modules in initramfs, and booting the instance without re-running dracut stalled when trying to load the disks. 

(Interestingly, dracut isn't required if the instance is created in vmware using `xfs` intead of `btrfs`. `btrfs` is the default filesystem on SLES 12.5)